### ISSUE TITLE
Breaking change in 9.0.1 for custom contract resolvers? 

### ISSUE THREAD

**Issue by ReeceWilliams** [JamesNK/Newtonsoft.Json#950]
We've a custom contract resolver inheriting from CamelCasePropertyNamesContractResolver in our ASP Web API :

```
    public class MyCamelCaseContractResolver : CamelCasePropertyNamesContractResolver
    {
        /// <summary>
        /// Extends ancestor class by converting trailing "ID" to "Id".
        /// </summary>
        protected override string ResolvePropertyName(string propertyName)
        {
            var result = base.ResolvePropertyName(propertyName);
            if (result.EndsWith("ID", false, CultureInfo.InvariantCulture))
            {
                return result.Replace("ID", "Id");
            }
            return result;
        }
        
    }
```

This is set in WebApiConfig:
```
            // Use camel case for all JSON data.
            config.Formatters.JsonFormatter.SerializerSettings = new JsonSerializerSettings
            {
                ContractResolver = new MyCamelCaseContractResolver()
            };
```

This all worked fine in 8.0.3, but after updating to 9.0.1 it no longer works - properties are returned as "xxxID" instead of "xxxId".

Even if we use it explicitly it still doesn't work:
```
JsonConvert.SerializeObject(data,
              JsonSerializerSettings
                       {
                            ContractResolver = new MyCamelCaseContractResolver()
                        })
```

If there's a good reason why this no longer works, is there a new way of doing this now?

Thanks,
Reece.

**Comment by JamesNK** [comment]
Yes that will now break. That method is never called because of changes in how CamelCasePropertyNamesContractResolver works.

What you could do is inherit from CamelCaseNamingStrategy and do something similar, then assign that to DefaultContractResolver. Note that you should cache the contract resolver on a static variable so it isn't recreated all the time.

**Comment by ReeceWilliams** [comment]
Cool, that worked but I also had to set ProcessDictionaryKeys to true on the strategy. 
Ended up with the following:
```
    public class MyCamelCaseContractResolver : CamelCasePropertyNamesContractResolver
    {
        public static readonly JsonSerializerSettings Settings = new JsonSerializerSettings
        {
            ContractResolver = new MyCamelCaseContractResolver()
        };

        public MyCamelCaseContractResolver()
        {
            this.NamingStrategy = new MyCamelCaseNamingStrategy();
        }
    }

    public class MyCamelCaseNamingStrategy : CamelCaseNamingStrategy
    {
        public MyCamelCaseNamingStrategy()
        {
            this.ProcessDictionaryKeys = true;
        }
        protected override string ResolvePropertyName(string name)
        {
            var result = base.ResolvePropertyName(name);
            if (result.EndsWith("ID", false, CultureInfo.InvariantCulture))
            {
                return result.Replace("ID", "Id");
            }
            return result;
        }
    }
```

Then...
```
// Use camel case for all JSON data.
config.Formatters.JsonFormatter.SerializerSettings = CSnetCamelCaseContractResolver.Settings;
```

**Comment by deanro** [comment]
I know I'm a little late to the party :-), but I just ran into this myself as I just upgraded our code base from version 6/7 to the latest. 

I found that I had to set this.OverrideSpecifiedNames to true in my custom NamingStrategy class. Setting ProcessDictionaryKeys to true didn't help.

It would have been nice if this breaking changes was mentioned on the release notes page https://github.com/JamesNK/Newtonsoft.Json/releases I did go through those yesterday and searched for breaking and didn't find anything. 

This post definitely helped though.
Thanks
Dean.
